File: packages/core/src/store/quiz-store.ts
import { create } from 'zustand';
import type { Session, Strategy } from '../types.js';
import { createSession, submitAnswer, nextQuestion, isSessionComplete, getSessionScore } from '../quiz-engine.js';
import type { Question } from '@aws-prep/content';

interface QuizState {
  session: Session | null;
  start: (questions: Question[], strategy: Strategy, examId?: number) => void;
  answer: (picked: string[], timeMs: number) => void;
  advance: () => void;
  reset: () => void;
  score: () => { correct: number; total: number; pct: number } | null;
  isComplete: () => boolean;
}

export const useQuizStore = create<QuizState>((set, get) => ({
  session: null,

  start: (questions, strategy, examId) => {
    set({ session: createSession(questions, strategy, examId) });
  },

  answer: (picked, timeMs) => {
    const { session } = get();
    if (!session) return;
    set({ session: submitAnswer(session, picked, timeMs) });
  },

  advance: () => {
    const { session } = get();
    if (!session) return;
    set({ session: nextQuestion(session) });
  },

  reset: () => set({ session: null }),

  score: () => {
    const { session } = get();
    return session ? getSessionScore(session) : null;
  },

  isComplete: () => {
    const { session } = get();
    return session ? isSessionComplete(session) : false;
  },
}));

File: myXcodeApp/myXcodeApp/Stores/QuizStore.swift
import Foundation
import Combine

@MainActor
final class QuizStore: ObservableObject {
    static let shared = QuizStore()

    @Published var session: QuizSession?
    @Published var selectedOptions: Set<String> = []
    @Published var submitted = false
    @Published var showTutor = false

    private let progress = ProgressStore.shared
    private let flashcards = FlashcardStore.shared
    private let startTime = { Date() }

    private var questionStartTime: Date = Date()

    private init() {}

    func startSession(questions: [ContentQuestion], strategy: QuizStrategy, examId: Int? = nil) {
        session = createSession(questions: questions, strategy: strategy, examId: examId)
        selectedOptions = []
        submitted = false
        questionStartTime = Date()
    }

    func toggleOption(_ letter: String, isMultiple: Bool) {
        guard !submitted else { return }
        if isMultiple {
            if selectedOptions.contains(letter) {
                selectedOptions.remove(letter)
            } else {
                selectedOptions.insert(letter)
            }
        } else {
            selectedOptions = [letter]
        }
    }

    func submit() {
        guard var s = session, !submitted, !selectedOptions.isEmpty else { return }
        let timeMs = Int(Date().timeIntervalSince(questionStartTime) * 1000)
        let picked = Array(selectedOptions).sorted()
        submitAnswer(session: &s, picked: picked, timeMs: timeMs)
        session = s
        submitted = true

        if let q = s.questions[safe: s.currentIdx] {
            let correct = gradeAnswer(question: q, picked: picked)
            progress.recordAnswer(examId: q.examId, questionNumber: q.number, picked: picked, correct: correct)
            if !correct {
                flashcards.addMistakeCard(question: q, picked: picked)
            }
        }
    }

    func advance() {
        guard var s = session else { return }
        advanceSession(session: &s)
        session = s
        selectedOptions = []
        submitted = false
        questionStartTime = Date()
    }

    func reset() {
        session = nil
        selectedOptions = []
        submitted = false
    }

    var currentQuestion: ContentQuestion? { session?.currentQuestion }
    var isComplete: Bool { session?.isComplete ?? false }
    var isMultipleAnswer: Bool { (currentQuestion?.correctLetters.count ?? 1) > 1 }
}

extension Array {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
